packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Models/Types/MrwSerializationTypeProvider.cs

```diff
@@ -124,30 +132,54 @@ protected override MethodProvider[] BuildConstructors()
 
             var FieldProvider = new FieldProvider(
                 modifiers: FieldModifiers.Private,
-                type: _privateAdditionalPropertiesPropertyType,
-                name: _privateAdditionalPropertiesPropertyName);
+                type: _privateAdditionalRawDataPropertyType,
+                name: _privateAdditionalRawDataPropertyName);
 
             return FieldProvider;
         }
 
+        private PropertySerializationProvider[] BuildPropertySerializations()
+        {
+            var propertyCount = _inputModel.Properties.Count;
+            PropertySerializationProvider[] propertySerializations = new PropertySerializationProvider[propertyCount];
+            for (var i = 0; i < propertyCount; i++)
+            {
+                var property = _inputModel.Properties[i];
+                propertySerializations[i] = new PropertySerializationProvider(property);
+            }
+
+            return propertySerializations;
+        }
+
         /// <summary>
         /// Builds the serialization methods for the model.
         /// </summary>
         /// <returns>A list of serialization and deserialization methods for the model.</returns>
         protected override MethodProvider[] BuildMethods()
         {
-            // TO-DO: Add deserialization methods https://github.com/microsoft/typespec/issues/3330
-
-            return new MethodProvider[]
+            var methods = new List<MethodProvider>();
+            // Add JSON serialization method declarations
+            if (_model.Inherits == null)
             {
-                // Add JSON serialization methods
-                BuildJsonModelWriteMethod(),
-                BuildJsonModelCreateMethod(),
-                // Add IModel methods
-                BuildIModelWriteMethod(),
-                BuildIModelCreateMethod(),
-                BuildIModelGetFormatFromOptionsMethod()
-            };
+                methods.Add(BuildJsonModelWriteMethodDeclaration());
+                methods.Add(BuildJsonModelCreateMethodDeclaration());
+                methods.Add(BuildIModelWriteMethodDeclaration());
+                methods.Add(BuildIModelCreateMethodDeclaration());
+            }
+
+            // Add JSON serialization methods
+            methods.Add(BuildJsonModelWriteMethod());
+            methods.Add(BuildJsonModelCreateMethod());
+            // Add Deserialization method
+            methods.Add(BuildDeserializationMethod());
+            // Add IModel methods
+            methods.Add(BuildIModelWriteMethod());
+            methods.Add(BuildIModelCreateMethod());
+            methods.Add(BuildIModelGetFormatFromOptionsMethod());
+            methods.Add(BuildFromOperationResponseMethod());
+            methods.Add(BuildToBinaryContentMethod());
```

Comment:
I think we need some comments explaining why when the model has a base type, it does some extra stuff, because we should always have 5 implemenation methods for the interface.

Reply:
Oh I see, those extra methods are `WriteCore` `CreateCore` etc.
I think the method names are not so clear, please add comments and/or refine those method names.